articles/quickstart/backend/java-spring-security5/01-authorization.md

```diff
@@ -0,0 +1,317 @@
+
+The example below shows how to secure API methods using the `HttpSecurity` object provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization required:
+
+```java
+// src/main/java/com/auth0/example/security/SecurityConfig.java
+
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Override
+    public void configure(HttpSecurity http) throws Exception {
+        http.authorizeRequests()
+                .mvcMatchers("/api/public").permitAll()
+                .mvcMatchers("/api/private").authenticated()
+                .mvcMatchers("api/private-scoped").hasAuthority("SCOPE_read:messages")
+                .and()
+                .oauth2ResourceServer().jwt();
+    }
+}
+```
+
+::: note
+By default, Spring Security will create a `GrantedAuthority` for each scope in the `scope` claim of the JWT. This is what enables using the `hasAuthority("SCOPE_read:messages")` method to restrict access to a valid JWT that contains the `read:messages` scope.
+
+If your use case requires different claims to make authorization decisions, see the [Spring Security Reference Documentation](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2resourceserver-authorization-extraction) for detailed instructions.
+:::
+
+### Create the API controller
+
+Create a new class named `Message`, which is the domain object the API will return:
+
+```java
+// src/main/java/com/auth0/example/model/Message.java
+
+public class Message {
+    private final String message;
+
+    Message(String message) {
+        this.message = message;
+    }
+
+    public String getMessage() {
+        return this.message;
+    }
+}
+```
+
+Create a new class named `APIController` to handle requests to the endpoints:
+
+```java
+// src/main/java/com/auth0/example/web/APIController.java
+
+@RestController
+@RequestMapping(path = "api", produces = MediaType.APPLICATION_JSON_VALUE)
+public class APIController {
+
+    @GetMapping(value = "/public")
+    public Message publicEndpoint() {
+        return new Message("All good. You DO NOT need to be authenticated to call /api/public.");
+    }
+
+    @GetMapping(value = "/private")
+    public Message privateEndpoint() {
+        return new Message("All good. You can see this because you are Authenticated.");
+    }
+
+    @GetMapping(value = "/private-scoped")
+    public Message privateScopedEndpoint() {
+        return new Message("All good. You can see this because you are Authenticated with a Token granted the 'read:messages' scope");
+    }
+}
+```
+
+## Run and Test Your API
+
+To build and run the sample project, run the `bootRun` Gradle task.
+
+Linux or macOS:
+```bash
+./gradlew bootRun
+```
+
+Windows:
+```bash
+gradlew.cmd bootRun
+```
+
+Using a REST client such as Postman or cURL, issue a `GET` request to `http://localhost:3010/api/public`. You should receive the response:
```

Comment:
All this is also explained on other snippets. Check out the backend includes folder for other things that you could reuse. e.g. for this one, it would be `_api_using.md`

https://github.com/auth0/docs/tree/master/articles/quickstart/backend/_includes

Reply:
The motivation for including this entire section in the first section of the Quickstart article was based on my own experience getting the Spring Security API sample running. The `_api_using.md` document contains information about obtaining and using access tokens in general, along with information that you mention for calling the different endpoints. Couple issues I had with this were:

* I didn't want to read through additional pages to know how to simply test what I just did.
* The Testing section in the `_api_using.md` document describe how to use the token to call the various endpoints, but not how to get one with the correct scopes.

All that said, there is some duplicate info as you mention. What if we:

* Update the `_api_using.md` with instructions on obtaining a test token with and without the proper permissions.
* Remove this section on testing your API from the "Authorization" page of the Quickstart article.
* Also place the instructions to get a test token in the `README` of the sample repo, for those who go right to the repo.

Some of this is related to two separate sources-of-truth (docs and sample), and I believe it's a goal for 2019 to fix that. But what do you think of the above proposal?

Reply:
Updated `_api_using.md` to include instructions for obtaining test tokens, and removed the Testing instructions from the Authorization article.